src/Hpack/Render.hs

```diff
@@ -140,11 +141,21 @@ renderPackage settings alignment existingFieldOrder sectionsFieldOrder Package{.
       executablesCabalVersion = foldr max Nothing . map executableCabalVersion . Map.elems
 
       executableCabalVersion :: Section Executable -> Maybe Version
-      executableCabalVersion sect = makeVersion [2,0] <$ guard (executableHasGeneratedModules sect)
+      executableCabalVersion sect = maximum [
+          makeVersion [2,0] <$ guard (executableHasGeneratedModules sect)
+        , makeVersion [2,2] <$ guard (hasCxxParams sect)
+        ]
+
 
       executableHasGeneratedModules :: Section Executable -> Bool
       executableHasGeneratedModules = any (not . null . executableGeneratedModules)
 
+      hasCxxParams sect = or [
+          (not . null . sectionCxxOptions) sect
+        , (not . null . sectionCxxSources) sect
```

Comment:
Does this take effect for `cxx-*` in conditionals?  You may need to use `any` here.